zrml/court/src/lib.rs

```diff
@@ -27,221 +27,1131 @@ extern crate alloc;
+                        Error::<T>::OnlyDrawnJurorsCanVote
+                    );
+                    (index, draws[index].clone())
+                }
+                None => return Err(Error::<T>::OnlyDrawnJurorsCanVote.into()),
+            };
+
+            let vote = Vote::Secret { secret: secret_vote };
+            draws[index] = Draw { juror: who.clone(), vote, ..draw };
+
+            <Draws<T>>::insert(market_id, draws);
+
+            Self::deposit_event(Event::JurorVoted { juror: who, market_id, secret: secret_vote });
+            Ok(())
+        }
+
+        /// Denounce a juror during the voting period for which the secret vote is known.
+        /// This is useful to punish the behaviour that jurors reveal
+        /// their secrets before the voting period ends.
+        /// A check of `secret_hash == hash(juror ++ outcome ++ salt)` is performed for validation.
+        ///
+        /// # Arguments
+        ///
+        /// - `market_id`: The identifier of the court.
+        /// - `juror`: The juror whose secret vote might be known.
+        /// - `outcome`: The raw vote outcome which should match with the secret of the juror.
+        /// - `salt`: The hash which is used to proof that the juror did reveal
+        /// her vote during the voting period.
+        ///
+        /// # Weight
+        ///
+        /// Complexity: `O(n)`, where `n` is the number of jurors
+        /// in the list of random selections (draws).
+        #[pallet::weight(1_000_000_000_000)]
+        #[transactional]
+        pub fn denounce_vote(
+            origin: OriginFor<T>,
+            #[pallet::compact] market_id: MarketIdOf<T>,
+            juror: AccountIdLookupOf<T>,
+            outcome: OutcomeReport,
+            salt: T::Hash,
+        ) -> DispatchResult {
+            let denouncer = ensure_signed(origin)?;
+
+            let juror = T::Lookup::lookup(juror)?;
+
+            ensure!(denouncer != juror, Error::<T>::SelfDenounceDisallowed);
+
+            let prev_juror_info = <Jurors<T>>::get(&juror).ok_or(Error::<T>::JurorDoesNotExists)?;
+
+            let court = <Courts<T>>::get(market_id).ok_or(Error::<T>::CourtNotFound)?;
+            let now = <frame_system::Pallet<T>>::block_number();
+            // ensure in vote period
+            ensure!(
+                court.periods.pre_vote_end < now && now <= court.periods.vote_end,
+                Error::<T>::NotInVotingPeriod
+            );
+
+            let mut draws = <Draws<T>>::get(market_id);
+            let (index, draw) = match draws.iter().position(|draw| draw.juror == juror) {
+                Some(index) => (index, draws[index].clone()),
+                None => return Err(Error::<T>::JurorNotDrawn.into()),
+            };
+
+            let secret = match draw.vote {
+                Vote::Secret { secret } => {
+                    ensure!(
+                        secret == T::Hashing::hash_of(&(juror.clone(), outcome.clone(), salt)),
```

Comment:
Please check if we can use the `T::Hashing::hash_of(&(juror.clone(), outcome.clone(), salt))` style.

Reply:
Does this function invocation make sense from a front end perspective? Is it clear from a caller perspective. So I didn't test it yet on the node side. But even if I did, we should check beforehand that it's also a valid way for the front end.

In other words. Can the front end generate a secret hash with those three components (juror, outcome, salt) that matches the output of `T::Hashing::hash_of(&(juror.clone(), outcome.clone(), salt))`?

Reply:
We could also use something like `(juror.clone(), outcome.clone(), salt).using_encoded(T::Hashing::hash)` or is this the same?

Reply:
7cbdbe5a55b7ec886f2537ff515506b36c069cec

Reply:
1c732de3dcd978bf31f39c064a54e2ae03990659